File: erubit.platform/src/main/java/a/erubit/platform/course/CharacterLesson.java
package a.erubit.platform.course;

import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;

import u.U;

public class CharacterLesson extends BunchLesson {
    private static final int RANK_FAMILIAR = 1;
    private static final int RANK_LEARNED = 2;
    private static final int RANK_LEARNED_WELL = 3;

    CharacterLesson(Course course) {
        super(course);
    }

    @Override
    protected int getRankFamiliar() {
        return RANK_FAMILIAR;
    }
    @Override
    protected int getRankLearned() {
        return RANK_LEARNED;
    }
    @Override
    protected int getRankLearnedWell() {
        return RANK_LEARNED_WELL;
    }


    @NonNull
    @Override
    protected PresentableDescriptor getPresentable(Item problemItem) {
        Random random = new Random();
        int apxSize = 4 + random.nextInt(4);

        String regex = ",\\s+";
        Problem problem = new Problem(this, problemItem);
        String[] words = U.defurigana(problem.meaning).split(regex);
        problem.variants = new String[words.length + apxSize];

        ArrayList<String> variants = new ArrayList<>(problem.variants.length);
        variants.addAll(Arrays.asList(words));
        for (int k = 0; k < apxSize; k++) {
            String[] randomWords = U.defurigana(mSet.get(random.nextInt(mSet.size())).meaning).split(regex);
            for (String s : randomWords) {
                if (!variants.contains(s))
                    variants.add(s);
            }
        }

        // add noise if has empty slots
        for (int k = variants.size(); k < problem.variants.length; k++)
            variants.add(mVariants.get(random.nextInt(mVariants.size())));

        problem.variants = variants.toArray(problem.variants);

        U.shuffleStrArray(problem.variants);

        return new PresentableDescriptor(problem);
    }
}

File: erubit.platform/src/main/java/a/erubit/platform/course/CharacterLesson.kt
package a.erubit.platform.course

import u.U
import java.util.*


class CharacterLesson internal constructor(course: Course) : BunchLesson(course) {
	override val rankFamiliar: Int
		get() = 1
	override val rankLearned: Int
		get() = 2
	override val rankLearnedWell: Int
		get() = 3

	override fun getPresentable(problemItem: Item): PresentableDescriptor {
		val random = Random()
		val apxSize = 4 + random.nextInt(4)
		val regex = "\\s+:\\s+"
		val problem = Problem(this, problemItem)
		val words = U.defurigana(problem.meaning).split(regex).toTypedArray()

		problem.variants = arrayOfNulls(words.size + apxSize)

		val variants = ArrayList<String>(problem.variants.size)
		variants.addAll(listOf(*words))

		val set = mSet ?: return PresentableDescriptor.ERROR
		val noise = mVariants ?: return PresentableDescriptor.ERROR

		for (k in 0 until apxSize) {
			val randomWords = U.defurigana(set[random.nextInt(set.size)].meaning).split(regex).toTypedArray()
			for (s in randomWords)
				if (!variants.contains(s))
					variants.add(s)
		}

		// add noise if has empty slots
		for (k in variants.size until problem.variants.size)
			variants.add(noise[random.nextInt(noise.size)])

		problem.variants = variants.toArray(problem.variants)
		U.shuffleStrArray(problem.variants)

		return PresentableDescriptor(problem)
	}
}
